### DISCUSSION TITLE
A Modal Portal in Lit

### DISCUSSION THREAD

**Discussion by nicholas-wilcox** [lit/lit#3041] [Show and tell]
A project that I am developing for my company is a single page application created with Lit components. The requirements involved being able to open the same modal from different components, or from the same component that can appear in different parts of the DOM. I effectively needed to separate the template of the modal from the component that causes it to appear. Rendering a template somewhere else in the DOM is precisely what React portals are for.

So, I did my best to create something that works like a React portal, but built in Lit. You can find the repository [here](https://github.com/cirrus-logic/lit-modal-portal), and test out the package [on NPM](https://www.npmjs.com/package/lit-modal-portal). It works by implementing a `<modal-portal>` custom element that you place at the bottom of your page, that attaches to a singleton controller that can be used to add/remove Lit templates from the "modal stack" (i.e. the DOM children of the `<modal-portal>`).

You can either use the `modalController.push()` method to programmatically push a single modal, or use a `portal()` directive that allows the modal to reactively update with the component that used the directive.

I don't really have the interest or bandwidth in actively maintaining the library, although I'll certainly take a look at any pull requests or bug reports. My manager and I just thought it would be a good thing to give back to the community, if only as an example that people can copy or modify to use in their own projects.

**Comment by nicholas-wilcox** [comment]
Also, not to embarrass myself, but this is my first time publishing something NPM. I _wanted_ to add a minified build that included our other dependency besides Lit, [Immutable.js](https://immutable-js.com/), which like `lit-modal-portal` is licensed under MIT. However, I couldn't not get the minified build (which is a JavaScript module) to work in a practical scenario without also bundling Lit. I have the feeling that this conflicts with Lit's BSD 3-Clause license. I just thought I'd mention it and defer to the Lit team.

**Reply by justinfagnani** [reply]
You usually shouldn't bundle a library. Let applications do that so that if they also use your dependencies they don't get them twice.

**Comment by justinfagnani** [comment]
Thanks for showing your library!

This is a pretty common need, and something we had addressed in Polymer with [IronOverlayBehavior](https://www.npmjs.com/package/@polymer/iron-overlay-behavior). We've been thinking we should probably port that (and a couple of Iron helpers) to Lit and make them available - which would be very much along the lines of this package here.

I didn't have time to look too deeply, but I think we'll look into what you did, and hopefully if we publish something you won't have to maintain this if you don't have the time :)

**Comment by keviddles** [comment]
Thanks for showing your work, @nicholas-wilcox! Very nice implementation.

From what I can tell of your implementation, Nicholas, your `portal` directive is manually adding / removing the modal from its internal registry whenever the boolean state changes, correct? And from what I can tell, you're not using slots anywhere, right?

---

I'm also exploring a portal implementation using Lit. Here is a StackBlitz:

https://stackblitz.com/edit/typescript-litelement-example-fugxhb?file=index.html,PortalMountPoint.ts,App.ts,Modal.ts,index.ts

In particular, I'm exploring whether it's possible to leverage `createRenderRoot` to enable creating Lit components that can exist within the render tree of a sub component, but actually get attached at a different mount point in the tree. So that we'd be able to do something like:

```
<body>
  <some-app>
    <some-page-in-my-app>
      <some-sub-component-that-contains-state>
        /* "some-sub-component" would conditionally render `my-modal` and pass in slotted content */
        <my-modal>Some content goes here</my-modal> 
      </some-sub-component-that-contains-state
    </some-page-in-my-app>
    /* this is where I'd like things mounted */
    <some-portal-mount-point></some-portal-mount-point>
  </some-app>
</body>
```

I've gotten pretty close to a working implementation, in that the following works:

- [x] Property attributes are appropriately passed to the modal and rendered (i.e., `value`)
- [x] Styling works, as the element I'm attaching to (`some-portal-mount-point`) is a LitElement offering up a `shadowRoot`
- [x] Events are emitted from the modal's rendering parent - I'm passing this down as an `emitter` and dispatching events specifically from there. (This doesn't work for more generic built-in events like a click, but it _would_ work for custom events emitted from the modal, like a close event.)

However, the following _doesn't_ work, and I'm not sure if these are possible to do if leveraging `createRenderRoot`:

- [ ] Slots don't seem to forward their values if the element is not mounted at the right spot in the DOM tree
- [ ] The modal doesn't seem to respect state; it opens fine, but when closed (e.g., no longer returned from the `render` method) it sticks around in the DOM

I'm not sure if these limitations are inherent in the implementation of `createRenderRoot`, or whether I'm missing something in my POC. Would love to hear from @justinfagnani if you have any thoughts.

**Reply by keviddles** [reply]
A gif might be nice!

The `<portal-mount-point />` starts off showing some default content, that then gets replaced when the modal is present. The `value` property successfully updates its content in the modal. And then there's a demonstration of an event being emitted correctly.

However, the slot content in the modal is not displayed in the modal (instead it's present in the DOM where it's defined) and the modal fails to close upon click on the overlay.

![2022-09-22 17 03 54](https://user-images.githubusercontent.com/94634901/191850754-e3930486-969b-4932-aabd-42c43f1b4133.gif)

**Reply by SravaniPolisetty94** [reply]
Hey @keviddles @justinfagnani @nicholas-wilcox Is there any solution for porting dialog out of the shadow roots and having reactivity intact with the slotted elements events. I’m facing almost the similar issue that @keviddles has mentioned in the stackblitz

**Reply by justinfagnani** [reply]
@SravaniPolisetty94 the modern way to do this would be to use the `<dialog>` element: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog

It has very good browser support now, going back to Chrome 37 and Safari 15.4 (which runs on iPhone 7 and 6s).

**Comment by nicholas-wilcox** [comment]
Bumping this because I recently refactored my library. I removed most of the code and focused on creating a custom `portal` directive that was simple to use and could support different use cases.

Also, shout out to the [lit-portal](https://github.com/ronak-lm/lit-portal) package made by @ronak-lm, which is a much better component-based solution than my package.

**Comment by fredericbahr** [comment]
@justinfagnani are there any plans to add Portals as a first-class support to Lit in the near future, maybe as a package within lit/labs?

**Reply by tbroyer** [reply]
https://github.com/lit/lit/discussions/3041#discussioncomment-14211844

**Reply by ronak-lm** [reply]
I guess it would be a technical limitation. Since lit only works in a shadow dom and portals (in React and others) find the <body /> (or the target tag) and attach themselves to it. Within the shadow dom, you can't technically find the body or the target.

The component I built was also kind of a hacky way to achieve it.

**Reply by justinfagnani** [reply]
Lit works with or without shadow DOM.

Shadow DOM isn't the issue though. You could easily build a Lit-based portal system. I have something like it for my own projects.

I think the questions are about how much it fits developer expectations, especially around styling since the portal might be out of the host component's tree, and whether it belongs in core.

For an example of how you can do a portal-like-thing, you can call `render()` on any element. If you provide a portal target via context, components can render to it. Or what I like to do is fire an event that carries a `render()` callback. The portal's owner container can listen for these events and call `render()` into the portal element.
